README.txt
Produce & Publish Lite
======================

Produce & Publish Lite is a stripped down version of the Produce & Publish
platform (www.produce-and-publish.com). It implements a PDF generation
functionality for the Plone content-management system (www.plone.org).

Limitations
-----------
* supports Apache FOP 1.0 or PISA as PDF backend
* no support for the Produce & Publish server (conversion will
  take place inside Plone - may block worker threads of Zope)

Requirements
------------
* requires Plone 4 or higher (untested with Plone 3.x)

Installation
------------

* add ``zopyx.smartprintng.lite`` to the ``eggs`` option
  of your buildout.cfg and re-run buildout (as usual)
* read carefully the ``zopyx.convert2`` documentation (especially the
  section related to the configuration of Apache FOP)

Installation of Apache using ``zc.buildout``
--------------------------------------------

You can automatically install and configure Apache FOP through
your buildout configuration::

    [buildout]
    parts = 
        ...
        fop

    [instance]
    ...
    environment-vars = 
        FOP_HOME ${buildout:directory}/parts/fop


    [fop]
    recipe = hexagonit.recipe.download
    strip-top-level-dir = true
    url = http://apache.openmirror.de/xmlgraphics/fop/binaries/fop-1.0-bin.tar.gz


Usage
-----
The module supports conversion to PDF either using the FOP converter or PISA
(installed automatically).  Add ``@@asPlainPDF`` to the URL of a Plone
document, news item or topic in order to convert the current content to PDF.
This is use the default PDF converter (Apache FOP). In order to convert the
HTML content using PISA you have to use ``@@asPlainPDF?converter=pdf-pisa``.


Supplementary information
-------------------------

PDF support for your own content-types:

    http://docs.produce-and-publish.com/connector/content-types.html

Registering your own resource directories:

    http://docs.produce-and-publish.com/connector/resource-directories.html#registering-your-own-resource-directory


License
-------
Produce & Publish Lite is published under the GNU Public License V 2 (GPL 2).


Author
------

| ZOPYX Limited
| c/o Andreas Jung
| Charlottenstr. 37/1
| D-72070 Tuebingen, Germany
| www.zopyx.com
| info@zopyx.com


docs/HISTORY.txt
Changelog
=========

1.0.6 (2011/06/15)
------------------
* no more generating a width=100% style for .image-container
  in order to avoid issues with PISA

1.0.5 (2011/06/05)
------------------
* supports PDF generation through Pisa (xhtml2pdf project)
  (internal pdf-pisa converter name)

1.0.4 (2011/06/04)
------------------
* documentation update

1.0.2 (2011/06/03)
------------------
* released with full source code

1.0.1 (2010/10/30)
------------------
* minor fix

1.0.0 (2010/10/20)
------------------
* minor fix

1.0.0b1 (2010/10/05)
--------------------
* initial release


zopyx/smartprintng/lite/browser/topic.py
################################################################
# zopyx.smartprintng.lite
# (C) 2009,  ZOPYX Limited & Co. KG, D-72070 Tuebingen, Germany
################################################################

from compatible import InitializeClass
from Products.Five.browser import BrowserView

try:
    from zope.app.pagetemplate.viewpagetemplatefile import ViewPageTemplateFile
except ImportError:
    from zope.browserpage.viewpagetemplatefile import ViewPageTemplateFile

class HTMLView(BrowserView):
    """ This view renders a HMTL fragment for the configured content type """

    def collect(self, html):

        html.append('<div class="type-topic">')
        for brain in self.context.queryCatalog():
            obj = brain.getObject()
            view = obj.restrictedTraverse('@@asHTML', None)
            if view:
                html.append('<div class="topic-item">')
                html.append(view())
                html.append('</div>')
            else:
                html.append('<div class="topic-item">')
                html.append('<span class="aggregation-error">new view for %s (%s) found</span>' %
                            (obj.absolute_url(1), obj.portal_type))
                html.append('</div>')

        html.append('</div>')

    def __call__(self, *args, **kw):
        """ Collector for topic content """
        html = list()
        self.collect(html)
        return ''.join(html)

InitializeClass(HTMLView)



zopyx/smartprintng/lite/browser/pdf.py
################################################################
# zopyx.smartprintng.lite
# (C) 2009,  ZOPYX Limited & Co. KG, D-72070 Tuebingen, Germany
################################################################

import os
import glob
import shutil
import tempfile
import urllib2
import BeautifulSoup

from compatible import InitializeClass
from Products.Five.browser import BrowserView
from Products.Five.browser.pagetemplatefile import ViewPageTemplateFile

try: 
    from zope.app.pagetemplate.viewpagetemplatefile import ViewPageTemplateFile as ViewPageTemplateFile2
except ImportError:
    from zope.browserpage.viewpagetemplatefile import ViewPageTemplateFile as ViewPageTemplateFile2

from zopyx.convert2 import Converter
from zopyx.smartprintng.lite.logger import LOG
from zopyx.smartprintng.lite.resources import resources_registry
from transformation import Transformer
from util import getLanguageForObject

cwd = os.path.dirname(os.path.abspath(__file__))


class PDFView(BrowserView):
    """ PDF view (using SmartPrintNG server) """

    template = ViewPageTemplateFile('resources/pdf_template.pt')
    resources = 'resources'

    # default transformations used for the default PDF view.
    # 'transformations' can be overriden within a derived PDFView.
    # If you don't need any transformation -> redefine 'transformations'
    # as empty list or tuple

    transformations = (
        'makeImagesLocal',
        )

    def copyResources(self, resources_dir, destdir):
        """ Copy over resources for a global or local resources directory into the 
            destination directory.
        """
        if os.path.exists(resources_dir):
            for name in os.listdir(resources_dir):
                fullname = os.path.join(resources_dir, name)
                if os.path.isfile(fullname):
                    shutil.copy(fullname, destdir)

    def transformHtml(self, html, destdir, transformations=None):
        """ Perform post-rendering HTML transformations """

        if transformations is None:
            transformations = self.transformations

        # the request can override transformations as well
        if self.request.has_key('transformations'):
            t_from_request = self.request['transformations']
            if isinstance(t_from_request, basestring):
                transformations = t_from_request and t_from_request.split(',') or []
            else:
                transformations = t_from_request

        T = Transformer(html, destdir, self.context)
        T(transformations)
        return T.render()

    def __call__(self, *args, **kw):

        try:
            return self.__call2__(*args, **kw)
        except:
            LOG.error('Conversion failed', exc_info=True)
            raise


    def __call2__(self, *args, **kw):
        """ URL parameters:
            'language' -  'de', 'en'....used to override the language of the
                          document
            'converter' - default to on the converters registered with
                          zopyx.convert2 (default: pdf-prince)
            'resource' - the name of a registered resource (directory)
            'template' - the name of a custom template name within the choosen
                         'resource' 
        """

        # Output directory
        destdir = tempfile.mkdtemp()

        # debug/logging
        params = kw.copy()
        params.update(self.request.form)
        LOG.debug('new job (%s, %s) - outdir: %s' % (args, params, destdir))

        # get hold of the language (hyphenation support)
        language = getLanguageForObject(self.context)
        if params.get('language'):
            language = params.get('language')

        # Check for CSS injection
        custom_css = None
        custom_stylesheet = params.get('custom_stylesheet')
        if custom_stylesheet:
            custom_css = str(self.context.restrictedTraverse(custom_stylesheet, None))
            if custom_css is None:
                raise ValueError('Could not access custom CSS at %s' % custom_stylesheet)

        # check for resource parameter
        resource = params.get('resource')
        if resource:
            resources_directory = resources_registry.get(resource)
            if not resources_directory:
                raise ValueError('No resource "%s" configured' % resource)
            if not os.path.exists(resources_directory):
                raise ValueError('Resource directory for resource "%s" does not exist' % resource)
            self.copyResources(resources_directory, destdir)

            # look up custom template in resources directory
            template_name = params.get('template', 'pdf_template')
            if not template_name.endswith('.pt'):
                template_name += '.pt'
            template_filename = os.path.join(resources_directory, template_name)
            if not os.path.exists(template_filename):
                raise IOError('No template found (%s)' % template_filename)
            template = ViewPageTemplateFile2(template_filename)

        else:
            template = self.template

        # call the dedicated @@asHTML on the top-level node. For a leaf document
        # this will return either a HTML fragment for a single document or @@asHTML
        # might be defined as an aggregator for a bunch of documents (e.g. if the
        # top-level is a folderish object

        html_fragment = ''
        html_view = self.context.restrictedTraverse('@@asHTML', None)
        if html_view:
            html_fragment = html_view()

        # arbitrary application data
        data = params.get('data', None)

        # Now render the complete HTML document
        html = template(self,
                        language=language,
                        request=self.request,
                        body=html_fragment,
                        custom_css=custom_css,
                        data=data,
                        )

        html = self.transformHtml(html, destdir)

        # hack to replace '&' with '&amp;'
        html = html.replace('& ', '&amp; ')

        # and store it in a dedicated working directory
        dest_filename = os.path.join(destdir, 'index.html')

        # inject BASE tag
        pos = html.lower().find('<head>')
        if pos == -1:
            raise RuntimeError('HTML does not contain a HEAD tag')
        html = html[:pos] + '<head><base href="%s"/>' % dest_filename + html[pos+6:]

        file(dest_filename, 'wb').write(html)

        # copy over global styles etc.
        resources_dir = os.path.join(cwd, 'resources')
        self.copyResources(resources_dir, destdir)

        # copy over language dependent hyphenation data
        if language:
            hyphen_file = os.path.join(resources_dir, 'hyphenation', language + '.hyp')
            if os.path.exists(hyphen_file):
                shutil.copy(hyphen_file, destdir)

            hyphen_css_file = os.path.join(resources_dir, 'languages', language + '.css')
            if os.path.exists(hyphen_css_file):
                shutil.copy(hyphen_css_file, destdir)

        # now copy over resources (of a derived view)
        self.copyResources(getattr(self, 'local_resources', ''), destdir)

        c = Converter(dest_filename)
        result = c(params.get('converter', 'pdf-fop'))
        if result['status'] != 0:
            raise RuntimeError('Error during PDF conversion (%r)' % result)
        pdf_file = result['output_filename']
        LOG.debug('Output file: %s' % pdf_file)
        return pdf_file

InitializeClass(PDFView)


class PDFDownloadView(PDFView):

    def __call__(self, *args, **kw):

        pdf_file = super(PDFDownloadView, self).__call__(*args, **kw)

        # return PDF over HTTP
        R = self.request.response
        R.setHeader('content-type', 'application/pdf')
        R.setHeader('content-disposition', 'attachment; filename="%s.pdf"' % self.context.getId())
        R.setHeader('content-length', os.stat(pdf_file)[6])
        R.setHeader('pragma', 'no-cache')
        R.setHeader('cache-control', 'no-cache')
        R.setHeader('Expires', 'Fri, 30 Oct 1998 14:19:41 GMT')
        R.setHeader('content-length', os.stat(pdf_file)[6])
        return file(pdf_file, 'rb').read()

InitializeClass(PDFDownloadView)

